File: 2016/fase1/sanduíche/sanduíche.py
def count_sandwich_ways(N, D, sandwich):
    total_ways = 0
    
    # Contando as maneiras de pegar uma sequência contínua de pedaços
    prefix_sum = [0] * (N + 1)
    for i in range(1, N + 1):
        prefix_sum[i] = prefix_sum[i - 1] + sandwich[i - 1]
    for i in range(1, N + 1):
        for j in range(i, N + 1):
            if prefix_sum[j] - prefix_sum[i - 1] == D:
                total_ways += 1
    
    # Contando as maneiras de pegar pedaços das extremidades
    suffix_sum = [0] * (N + 1)
    for i in range(N, 0, -1):
        suffix_sum[i] = suffix_sum[i + 1] + sandwich[i - 1]
    for i in range(1, N):
        for j in range(i + 1, N + 1):
            if prefix_sum[i - 1] + suffix_sum[j] == D:
                total_ways += 1
    
    return total_ways

# Entrada de dados
N, D = map(int, input().split())
sandwich = list(map(int, input().split()))

# Chamada da função e saída
print(count_sandwich_ways(N, D, sandwich))

File: 2016/fase1/sanduíche/sanduíche.js
function contarFormasDeComerSanduiche(N, D, comprimentos) {
    let formas1 = 0;
    let formas2 = 0;

    // Contagem de formas com sequência contínua de pedaços
    let soma = 0;
    for (let i = 0; i < N; i++) {
        soma += comprimentos[i];
        if (soma === D) {
            formas1++;
        }
        if (soma > D) {
            break;
        }
    }

    // Contagem de formas com pedaços das extremidades
    let esquerda = 0;
    let direita = 0;
    soma = 0;
    for (let i = 0; i < N - 1; i++) {
        soma += comprimentos[i];
        while (soma > D && esquerda < i) {
            soma -= comprimentos[esquerda];
            esquerda++;
        }
        if (soma === D) {
            formas2 += i - esquerda + 1;
        }
    }

    return formas1 + formas2;
}

// Entrada
const entrada = "5 8\n1 2 3 2 1\n";
const linhas = entrada.trim().split("\n");
const [N, D] = linhas[0].split(" ").map(Number);
const comprimentos = linhas[1].split(" ").map(Number);

// Saída
console.log(contarFormasDeComerSanduiche(N, D, comprimentos));